python/mxnet/gluon/data/dataset.py

```diff
@@ -64,6 +64,37 @@ def filter(self, fn):
         from . import FilterSampler
         return _SampledDataset(self, FilterSampler(fn, self))
 
+    def shard(self, num_shards, index):
+        """Returns a new dataset includes only 1/num_shards of this dataset.
+
+        For distributed training, be sure to shard before you randomize the dataset
+        (such as shuffle), if you want each worker to reach a unique subset.
+
+        Parameters
+        ----------
+        num_shards : int
+            A integer representing the number of data shards.
+        index : int
+            A integer representing the index of the current shard.
+
+        Returns
+        -------
+        Dataset
+            The result dataset.
+        """
+        assert index < num_shards, 'Shard index of out bound: %d out of %d'%(index, num_shards)
+        assert num_shards > 0, 'Number of shards must be greater than 0'
+        assert index >= 0, 'Index must be non-negative'
+        length = len(self)
+        shard_len = length // num_shards
+        rest = length % num_shards
+        # Compute the start index for this partition
+        start = shard_len * index + min(index, rest)
```

Comment:
It will distribute the samples more uniformly.
For example, when there are 6 samples and 4 partitions, the partitions will get 2, 2, 1, 1 samples respectively.

Reply:
Got it, so it's assuming residules should be padded to early shards rather than latter.

Reply:
Yes.